sklearn/impute.py

```diff
@@ -40,6 +40,17 @@
 ]
 
 
+def _check_inputs_dtype(X, missing_values):
+    """Check that the dtype of X is in accordance with the one of
+    missing_values."""
+    if (X.dtype.kind in ("f", "i", "u") and
+            not isinstance(missing_values, numbers.Real)):
+        raise TypeError("The data type of 'missing_values' and 'X' are "
+                        "not compatible. 'missing_values' data type is "
+                        "{} and 'X' is {}."
```

Comment:
In the rest of the code, we raise a ValueError for this kind of error, there is a discussion in #11211.